overlord/hookstate/ctlcmd/fde_setup.go

```
@@ -0,0 +1,160 @@
+)
+
+type fdeSetupRequestCommand struct {
+	baseCommand
+}
+
+var shortFdeSetupRequestHelp = i18n.G("Request setup of full disk encryption")
+var longFdeSetupRequestHelp = i18n.G(`
+The fde-setup-request command is used inside the fde-setup hook. It will
+return information about what operation for full-disk encryption is
+requested and auxiliary data to complete this operation.
+
+The fde-setup hook should do what is requested and then call
+"snapctl fde-setup-result" and pass the result data to stdin.
+
+Here is an example for how the fde-setup hook is called initially:
+$ snapctl fde-setup-request
+{"op":"features"}
+$ echo '[]' | snapctl fde-setup-result
+
+Alternatively the hook could reply with:
+$ echo '{"error":"hardware-unsupported"}' | snapctl fde-setup-result
+
+And then it is called again with a request to do the initial key setup:
+$ snapctl fde-setup-request
+{"op":"initial-setup", "key": "key-to-seal", "key-name":"key-for-ubuntu-data"}
+$ echo "$sealed_key" | snapctl fde-setup-result
+`)
+
+func init() {
+	addCommand("fde-setup-request", shortFdeSetupRequestHelp, longFdeSetupRequestHelp, func() command { return &fdeSetupRequestCommand{} })
+}
+
+type fdeSetupJSON struct {
+	// XXX: make "op" a type: "features", "initial-setup", "update" ?
+	Op string `json:"op"`
+
+	Key     []byte `json:"key,omitempty"`
+	KeyName string `json:"key-name,omitempty"`
+
+	// Model related fields, this will be set to follow the
+	// secboot:SnapModel interface.
+	//
+	// XXX: do we need this to be a list? i.e. multiple models?
+	Model map[string]string `json:"model,omitempty"`
+
+	// TODO: provide LoadChains, KernelCmdline etc to support full
+	//       tpm sealing
+}
+
+func (c *fdeSetupRequestCommand) Execute(args []string) error {
+	context := c.context()
+	if context == nil {
+		return fmt.Errorf("cannot  without a context")
+	}
+	context.Lock()
+	defer context.Unlock()
+
+	var js fdeSetupJSON
+	if err := context.Get("fde-op", &js.Op); err != nil {
```

if err is ErrNoState we probably should return a clearer error about this not being the right kind of hook?

---

fde-op  called "fde-setup-op" could be the full FDESetupOP struct ?